app/validators/routing/types.py

```diff
@@ -42,33 +42,67 @@
     "NoneType": TYPE_NULL,
 }
 
+METADATA_TYPE_TO_JSON_TYPE = {
+    "string": TYPE_STRING,
+    "date": TYPE_STRING,
+    "boolean": TYPE_BOOLEAN,
+}
+
+
+def resolve_answer_source_json_type(value_source, schema):
+    answer_id = value_source["identifier"]
+    answer_type = schema.answers_with_context[answer_id]["answer"]["type"]
+    return ANSWER_TYPE_TO_JSON_TYPE[answer_type]
+
+
+def resolve_calculated_summary_source_json_type(value_source, schema):
+    block = schema.get_block(value_source["identifier"])
+    if block["calculation"].get("answers_to_calculate"):
+        answer_id = block["calculation"]["answers_to_calculate"][0]
+    else:
+        answer_value_source = block["calculation"]["operation"]["+"][0]
+        answer_id = answer_value_source["identifier"]
+    answer_type = schema.answers_with_context[answer_id]["answer"]["type"]
+    return ANSWER_TYPE_TO_JSON_TYPE[answer_type]
+
+
+def resolve_grand_calculated_summary_source_json_type(value_source, schema):
+    block = schema.get_block(value_source["identifier"])
+    first_calculated_summary_source = block["calculation"]["operation"]["+"][0]
+    return resolve_value_source_json_type(first_calculated_summary_source, schema)
+
+
+def resolve_metadata_source_json_type(value_source, schema):
+    if metadata_id := value_source.get("identifier"):
+        if metadata := schema.schema.get("metadata"):
+            for values in metadata:
+                if values.get("name") == metadata_id:
+                    return METADATA_TYPE_TO_JSON_TYPE[values.get("type")]
+    return TYPE_STRING
+
+
+def resolve_list_source_json_type(value_source):
+    if selector := value_source.get("selector"):
+        return LIST_SELECTOR_TO_JSON_TYPE[selector]
+    return TYPE_ARRAY
+
 
 def resolve_value_source_json_type(value_source, schema):
     source = value_source["source"]
```

Comment:
I would pull out both the source _and the identifier_ at this stage and then pass the identifier to the functions, saves you doing it in each one
(+ each of your new functions only needs the identifier, so nice to not pass in the whole value source)